crates/wasmparser/src/binary_reader.rs

```diff
@@ -242,78 +300,458 @@ impl<'a> BinaryReader<'a> {
+                    self.original_position() - 1,
+                ))
+            }
+        })
+    }
+
+    pub(crate) fn read_export(&mut self) -> Result<Export<'a>> {
+        Ok(Export {
+            name: self.read_string()?,
+            kind: self.read_external_kind()?,
+            index: self.read_var_u32()?,
+        })
+    }
+
+    pub(crate) fn read_component_export(&mut self) -> Result<ComponentExport<'a>> {
+        let name = self.read_string()?;
+        Ok(match self.read_u8()? {
+            0x00 => {
+                let offset = self.original_position();
+                let kind = self.read_u8()?;
+                let index = self.read_var_u32()?;
+                let kind = match kind {
+                    0x00 => ComponentExportKind::Module(index),
+                    0x01 => ComponentExportKind::Component(index),
+                    0x02 => ComponentExportKind::Instance(index),
+                    0x03 => ComponentExportKind::Function(index),
+                    0x04 => ComponentExportKind::Value(index),
+                    x => {
+                        return Err(BinaryReaderError::new(
+                            format!("invalid leading byte (0x{:x}) for export kind", x),
+                            offset,
+                        ))
+                    }
+                };
+                ComponentExport { name, kind }
+            }
+            0x01 => ComponentExport {
+                name,
+                kind: ComponentExportKind::Exports(
+                    (0..self.read_var_u32()?)
+                        .map(|_| self.read_component_export())
+                        .collect::<Result<_>>()?,
+                ),
+            },
+            x => {
+                return Err(BinaryReaderError::new(
+                    format!("invalid leading byte (0x{:x}) for export", x),
+                    self.original_position() - 1,
+                ))
+            }
+        })
     }
 
     pub(crate) fn read_import(&mut self) -> Result<Import<'a>> {
-        let module = self.read_string()?;
-        let field = self.read_string()?;
-
-        // For the `field`, figure out if we're the experimental encoding of
-        // single-level imports for the module linking proposal (a single-byte
-        // string which is 0xc0, which is invalid utf-8) or if we have a second
-        // level of import.
-        let field = if field.is_empty() && self.buffer.get(self.position) == Some(&0xff) {
-            self.position += 1;
-            None
-        } else {
-            Some(field)
-        };
+        Ok(Import {
+            module: self.read_string()?,
+            name: self.read_string()?,
+            ty: self.read_type_ref()?,
+        })
+    }
+
+    pub(crate) fn read_component_import(&mut self) -> Result<ComponentImport<'a>> {
+        Ok(ComponentImport {
+            name: self.read_string()?,
+            ty: self.read_var_u32()?,
```

Comment:
I think it's primarily based on the fact that some types in core wasm are "inlined" into the import section (basically anything that isn't a function) and components always refer to the import from the type section.

I should also note that per my feedback, Luke updated the spec today such that there's validation on these type indexes to ensure it can be any type other than a compound type definition. I haven't yet sync'd to those changes.

Reply:
Could this be updated to be a component-type-def in the binary format instead of something in the type section? I'm not sure if that would cause too much confusion in other places but naively it seems like we could do the same thing and it means you're not necessarily required to fill in the type section with something you just import and later use.